lib/pg_ha_migrations/safe_statements.rb

```diff
@@ -170,6 +170,65 @@ def safe_remove_concurrent_index(table, options={})
     unsafe_remove_index(table, **options.merge(:algorithm => :concurrently))
   end
 
+  def safe_add_concurrent_partitioned_index(table, columns, name_suffix: nil, if_not_exists: false)
+    if ActiveRecord::Base.connection.postgresql_version < 11_00_00
+      raise PgHaMigrations::InvalidMigrationError, "Concurrent partitioned index creation not supported on Postgres databases before version 11"
+    end
+
+    schema, table = _schema_and_table_for(table)
+
+    is_partitioned = select_value(<<~SQL)
+      SELECT EXISTS (
+        SELECT 1
+        FROM pg_partitioned_table, pg_class, pg_namespace
+        WHERE pg_class.oid = pg_partitioned_table.partrelid
+          AND pg_class.relnamespace = pg_namespace.oid
+          AND pg_class.relname = #{quote(table)}
+          AND pg_namespace.nspname = #{quote(schema)}
+      )
+    SQL
+
+    raise PgHaMigrations::InvalidMigrationError, "Table #{table} is not a partitioned table" unless is_partitioned
+
+    name_suffix = "#{Array.wrap(columns).join("_")}_idx" unless name_suffix.present?
+
+    child_tables = select_values(<<~SQL)
+      SELECT child.relname
+      FROM pg_inherits
+        JOIN pg_class parent ON pg_inherits.inhparent = parent.oid
+        JOIN pg_class child  ON pg_inherits.inhrelid  = child.oid
+        JOIN pg_namespace    ON parent.relnamespace = pg_namespace.oid
+      WHERE parent.relname = #{quote(table)}
+        AND pg_namespace.nspname = #{quote(schema)}
+    SQL
+
+    parent_index = "#{table}_#{name_suffix}"
+
+    quoted_columns = Array.wrap(columns).map { |c| quote_column_name(c) }.join(", ")
+
+    unsafe_execute(<<~SQL)
+      CREATE INDEX#{if_not_exists ? " IF NOT EXISTS" : ""} #{quote_column_name(parent_index)}
+      ON ONLY #{quote_schema_name(schema)}.#{quote_table_name(table)} (#{quoted_columns})
+    SQL
```

Comment:
I think probably this one needs a `say_with_time`, but the other ones we should just allow the output from `safe_add_concurrent_index` to come through.

Reply:
In addition to the `safe_add_concurrent_index` calls, we also have calls to `ALTER INDEX ... ATTACH PARTITION` though

Reply:
For those curious, this takes out a `ShareLock` on the parent table, just like any other `CREATE INDEX` operation (without `CONCURRENTLY`). This should be crazy fast though since it's not actually building the index, so I _don't think_ we need a call to `safe_acquire_lock_for_table` here

Reply:
We spoke offline, but for context:

This cannot be done with `CONCURRENTLY`, but it feels like maybe it should. We should hop on the mailing list to see if there's a technical reason preventing this, and if not make a contribution to postgres core